RetriesExhaustedException thrown by Client due to SocketTimeoutExceptions.

**Problem description**
RetriesExhaustedException thrown by Client due to SocketTimeoutExceptions.
```
2021-02-04 08:10:19,720 [Legacy Source Thread - Source: input-source -> Sink: input-source-print (1/1)] ERROR io.pravega.connectors.flink.FlinkPravegaReader               [] - Exception occurred while creating a Pravega EventStreamReader to read events
java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
	at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:375) ~[?:1.8.0_272]
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947) ~[?:1.8.0_272]
	at io.pravega.client.segment.impl.SegmentInputStreamImpl.handleRequest(SegmentInputStreamImpl.java:141) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.segment.impl.SegmentInputStreamImpl.read(SegmentInputStreamImpl.java:126) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.segment.impl.EventSegmentReaderImpl.readEvent(EventSegmentReaderImpl.java:93) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.segment.impl.EventSegmentReaderImpl.read(EventSegmentReaderImpl.java:74) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:154) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:125) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.connectors.flink.FlinkPravegaReader.run(FlinkPravegaReader.java:297) [blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:100) [flink-dist_2.12-1.11.2-spring1.jar:1.11.2-spring1]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:63) [flink-dist_2.12-1.11.2-spring1.jar:1.11.2-spring1]
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:213) [flink-dist_2.12-1.11.2-spring1.jar:1.11.2-spring1]
Caused by: io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
	at io.pravega.common.util.Retry$RetryAndThrowBase.lambda$runAsync$6(Retry.java:245) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:884) ~[?:1.8.0_272]
	at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:866) ~[?:1.8.0_272]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_272]
```

The rootcause for this is a SocketTTimeoutException

```
2021-02-04 08:10:19,504 [ClientSocketReaders-2369] WARN  io.pravega.client.connection.impl.TcpClientConnection        [] - Error processing data from from server nautilus-pravega-segment-store-1.pravega.kubespray.nautilus-platform-dev.com:12345
java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method) ~[?:1.8.0_272]
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116) ~[?:1.8.0_272]
	at java.net.SocketInputStream.read(SocketInputStream.java:171) ~[?:1.8.0_272]
	at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[?:1.8.0_272]
	at java.net.SocketInputStream.read(SocketInputStream.java:127) ~[?:1.8.0_272]
	at io.pravega.client.connection.impl.IoBuffer.getBuffOfSize(IoBuffer.java:61) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.connection.impl.TcpClientConnection$ConnectionReader.readCommand(TcpClientConnection.java:144) ~[blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
	at io.pravega.client.connection.impl.TcpClientConnection$ConnectionReader.run(TcpClientConnection.java:115) [blob_p-7cb7945ddab15bce79d51ee1ee85f1a8d34885b4-f57ef0e3e7535e7ae252a19084cea223:?]
```

Retry logs
```
Line 674: 2021-02-04 07:56:39,043 [ClientSocketReaders-2251] WARN  io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl   [] - Exception while reading from Segment video/adas-raw-video/2.#epoch.0 at offset 0 :
Line 1922: 2021-02-04 07:59:49,116 [ClientSocketReaders-2272] WARN  io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl   [] - Exception while reading from Segment video/adas-raw-video/2.#epoch.0 at offset 0 :
Line 3579: 2021-02-04 08:03:19,154 [ClientSocketReaders-2294] WARN  io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl   [] - Exception while reading from Segment video/adas-raw-video/2.#epoch.0 at offset 0 :
Line 5273: 2021-02-04 08:06:49,405 [ClientSocketReaders-2332] WARN  io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl   [] - Exception while reading from Segment video/adas-raw-video/2.#epoch.0 at offset 0 :
Line 6967: 2021-02-04 08:10:19,504 [ClientSocketReaders-2369] WARN  io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl   [] - Exception while reading from Segment video/adas-raw-video/2.#epoch.0 at offset 0 :
```

**Problem location**
KeepAlive mechanism between client and segmentStore.

**Suggestions for an improvement**
Fix it.

--- a/client/src/main/java/io/pravega/client/connection/impl/FlowHandler.java
+++ b/client/src/main/java/io/pravega/client/connection/impl/FlowHandler.java
@@ -253,8 +253,8 @@ public class FlowHandler extends FailingReplyProcessor implements AutoCloseable 
     private ReplyProcessor getReplyProcessor(Reply cmd) {
         int flowId = disableFlow.get() ? FLOW_DISABLED : Flow.toFlowID(cmd.getRequestId());
         final ReplyProcessor processor = flowIdReplyProcessorMap.get(flowId);
-        if (processor == null ) {
-            log.warn("No ReplyProcessor found for the provided flowId {} and reply {}. Ignoring response", flowId, cmd);
+        if (processor == null) {
+            log.warn("No ReplyProcessor found for the provided flowId {}. Ignoring response", flowId);
             if (cmd instanceof WireCommands.ReleasableCommand) {
                 ((WireCommands.ReleasableCommand) cmd).release();
             }

--- a/client/src/main/java/io/pravega/client/connection/impl/FlowHandler.java
+++ b/client/src/main/java/io/pravega/client/connection/impl/FlowHandler.java
@@ -147,6 +147,12 @@ public class FlowHandler extends FailingReplyProcessor implements AutoCloseable 
             return;
         }
 
+        if (cmd instanceof WireCommands.KeepAlive) {
+            // The SegmentStore responds to a KeepAlive WireCommand, this ensures the client can detect unresponsive
+            // SegmentStores due to network glitches. No action is required for the client on receiving this WireCommand.
+            return;
+        }
+
         // Obtain ReplyProcessor and process the reply.
         ReplyProcessor processor = getReplyProcessor(cmd);
         if (processor != null) {
@@ -247,8 +253,8 @@ public class FlowHandler extends FailingReplyProcessor implements AutoCloseable 
     private ReplyProcessor getReplyProcessor(Reply cmd) {
         int flowId = disableFlow.get() ? FLOW_DISABLED : Flow.toFlowID(cmd.getRequestId());
         final ReplyProcessor processor = flowIdReplyProcessorMap.get(flowId);
-        if (processor == null) {
-            log.warn("No ReplyProcessor found for the provided flowId {}. Ignoring response", flowId);
+        if (processor == null ) {
+            log.warn("No ReplyProcessor found for the provided flowId {} and reply {}. Ignoring response", flowId, cmd);
             if (cmd instanceof WireCommands.ReleasableCommand) {
                 ((WireCommands.ReleasableCommand) cmd).release();
             }

--- a/client/src/test/java/io/pravega/client/connection/impl/FlowHandlerTest.java
+++ b/client/src/test/java/io/pravega/client/connection/impl/FlowHandlerTest.java
@@ -59,6 +59,7 @@ import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
 import static org.mockito.Mockito.when;
 
 @RunWith(MockitoJUnitRunner.class)
@@ -229,6 +230,14 @@ public class FlowHandlerTest {
         flowHandler.process(msg);
         verify(processor).hello(msg);
         verify(errorProcessor).hello(msg);
+    }
+
+    @Test
+    public void testKeepAlive() {
+        final WireCommands.KeepAlive msg = new WireCommands.KeepAlive();
+        flowHandler.process(msg);
+        // ensure none of the Replyprocessors are bothered with this msg.
+        verifyNoInteractions(processor);
     }
 
     @Test

--- a/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java
+++ b/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java
@@ -130,6 +130,12 @@ public class AppendProcessor extends DelegatingRequestProcessor {
         }
     }
 
+    @Override
+    public void keepAlive(WireCommands.KeepAlive keepAlive) {
+        log.debug("Received a keepAlive from connection: {}", connection);
+        connection.send(keepAlive);
+    }
+
     /**
      * Setup an append so that subsequent append calls can occur.
      * This requires validating that the segment exists.

--- a/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/AppendProcessorTest.java
+++ b/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/AppendProcessorTest.java
@@ -106,6 +106,22 @@ public class AppendProcessorTest extends ThreadPooledTestSuite {
     }
 
     @Test
+    public void testKeepAlive() {
+        ServerConnection connection = mock(ServerConnection.class);
+        AppendProcessor processor = AppendProcessor.defaultBuilder()
+                .store(mock(StreamSegmentStore.class))
+                .connection(connection)
+                .connectionTracker(mock(ConnectionTracker.class))
+                .statsRecorder(Mockito.mock(SegmentStatsRecorder.class))
+                .build();
+
+
+        WireCommands.KeepAlive keepAliveCommand = new WireCommands.KeepAlive();
+        processor.keepAlive(keepAliveCommand);
+        verify(connection).send(new WireCommands.KeepAlive());
+    }
+
+    @Test
     public void testAppend() {
         String streamSegmentName = "scope/stream/0.#epoch.0";
         UUID clientId = UUID.randomUUID();

--- a/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/AppendProcessorTest.java
+++ b/segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/AppendProcessorTest.java
@@ -115,7 +115,6 @@ public class AppendProcessorTest extends ThreadPooledTestSuite {
                 .statsRecorder(Mockito.mock(SegmentStatsRecorder.class))
                 .build();
 
-
         WireCommands.KeepAlive keepAliveCommand = new WireCommands.KeepAlive();
         processor.keepAlive(keepAliveCommand);
         verify(connection).send(new WireCommands.KeepAlive());

--- a/shared/protocol/src/main/java/io/pravega/shared/protocol/netty/FailingReplyProcessor.java
+++ b/shared/protocol/src/main/java/io/pravega/shared/protocol/netty/FailingReplyProcessor.java
@@ -177,7 +177,7 @@ public abstract class FailingReplyProcessor implements ReplyProcessor {
 
     @Override
     public void keepAlive(KeepAlive keepAlive) {
-        throw new IllegalStateException("Unexpected operation: " + keepAlive);
+        log.info("KeepAlive received");
     }
 
     @Override

--- a/shared/protocol/src/main/java/io/pravega/shared/protocol/netty/FailingReplyProcessor.java
+++ b/shared/protocol/src/main/java/io/pravega/shared/protocol/netty/FailingReplyProcessor.java
@@ -177,7 +177,7 @@ public abstract class FailingReplyProcessor implements ReplyProcessor {
 
     @Override
     public void keepAlive(KeepAlive keepAlive) {
-        log.info("KeepAlive received");
+        log.trace("KeepAlive received");
     }
 
     @Override

--- a/shared/protocol/src/test/java/io/pravega/shared/protocol/netty/FailingReplyProcessorTest.java
+++ b/shared/protocol/src/test/java/io/pravega/shared/protocol/netty/FailingReplyProcessorTest.java
@@ -14,7 +14,6 @@ import io.pravega.shared.protocol.netty.WireCommands.AuthTokenCheckFailed;
 import io.pravega.shared.protocol.netty.WireCommands.ConditionalCheckFailed;
 import io.pravega.shared.protocol.netty.WireCommands.DataAppended;
 import io.pravega.shared.protocol.netty.WireCommands.InvalidEventNumber;
-import io.pravega.shared.protocol.netty.WireCommands.KeepAlive;
 import io.pravega.shared.protocol.netty.WireCommands.NoSuchSegment;
 import io.pravega.shared.protocol.netty.WireCommands.OperationUnsupported;
 import io.pravega.shared.protocol.netty.WireCommands.SegmentAlreadyExists;
@@ -64,7 +63,6 @@ public class FailingReplyProcessorTest {
         assertThrows(IllegalStateException.class, () -> rp.conditionalCheckFailed(new ConditionalCheckFailed(null, 1, 2)));
         assertThrows(IllegalStateException.class, () -> rp.dataAppended(new DataAppended(1, null, 0, -1, 2)));
         assertThrows(IllegalStateException.class, () -> rp.invalidEventNumber(new InvalidEventNumber(null, 0, "")));
-        assertThrows(IllegalStateException.class, () -> rp.keepAlive(new KeepAlive()));
         assertThrows(IllegalStateException.class, () -> rp.noSuchSegment(new NoSuchSegment(0, "", "", 2)));
         assertThrows(UnsupportedOperationException.class, () -> rp.operationUnsupported(new OperationUnsupported(0, "", "")));
         assertThrows(IllegalStateException.class, () -> rp.segmentAlreadyExists(new SegmentAlreadyExists(1, "", "")));

